### PR TITLE
Index to data/new and safe rotate, build script cleanup

### PR DESCRIPTION
**Pull request by zas**

### CODE REVIEW

File: bin/build [hunk 1]
```diff
@@ -62,8 +62,9 @@ if [ "$FREEDB_ELAPSED" -ge "$SEARCH_FREEDB_THRESHOLD" ] &&
    [ "$HOUR" -ge "$SEARCH_FREEDB_HOUR" ] &&
    [ "$HOUR" -le "$SEARCH_FREEDB_MAX_HOUR" ]; then
 	echo "Building freedb index. Last index was started $FREEDB_ELAPSED seconds ago"
-	date +%s > $FREEDB_TIMESTAMP_FILE
-	run_and_show_errors make freedb
+	START=$(date +%s)
+	run_and_show_errors make freedb || exit 2
+	echo "$START" > $FREEDB_TIMESTAMP_FILE
```
Comment (mayhem) [inline bin/build]: Ahhh, good!

File: bin/safe_dir_rotation [hunk 2]
```diff
@@ -0,0 +1,49 @@
+#!/bin/bash
+
```
Comment (mayhem) [inline bin/safe_dir_rotation:2]: Maybe add a usage statement or a couple of comments on how this should be invoked?

**Comment by mayhem** [comment]
I think this makes sense to me. However, I am unclear what you are trying to solve here -- what problem does this fix?

**Comment by zas** [comment]
[**@mayhem**](https://github.com/mayhem) commented on [5 déc. 2015 12:00 UTC+1](https://github.com/metabrainz/search-indexer/pull/1#issuecomment-162172651):

> I think this makes sense to me. However, I am unclear what you are trying to solve here -- what problem does this fix?

First, thanks for taking time to review this stuff.

Short explanation:
the whole process is not failure-resistant at all, we can't run indexer + search on the same machine easily with current scripts

Long explanation:

We have two way to rotate dirs:
- one is used in `build` script https://github.com/metabrainz/search-indexer/blob/d82fa5202836d2201de0e5b54a9c76fe559bc5b6/bin/build#L110-L122
- one is used by https://github.com/metabrainz/search-indexer/blob/d82fa5202836d2201de0e5b54a9c76fe559bc5b6/bin/receive-from-server#L31 based on https://github.com/metabrainz/search-indexer/blob/d82fa5202836d2201de0e5b54a9c76fe559bc5b6/bin/smart-rotate.py

**So what i want to "fix":**
- https://github.com/metabrainz/search-indexer/blob/d82fa5202836d2201de0e5b54a9c76fe559bc5b6/Makefile#L27 sets indexer write directory for new indexes to `./data/cur`, it would make much more sense to write to `./data/new`. Ie. if we want to dissociate index updates and index pushes (it could be two processes, one looping on index generation, the other syncing `./data/cur` after it was updated to a remote `./data/new`
- imagine indexer and servlet running on the same machine with current code: indexer writes to `./data/cur`, servlet reads from `./data/cur` -> boum !
- ensure potentially incomplete data is always written in `./data/new`, both on indexer (command from the `Makefile`), and the search (servlet) server (transfer using `netcat` + `tar`, in `receive-from-server`)
- when writing is ending successfully, do the rotation (new -> cur -> old). Note that current code fails on this, it potentially rotates even after a failure: 
  - if you have an error at https://github.com/metabrainz/search-indexer/blob/d82fa5202836d2201de0e5b54a9c76fe559bc5b6/bin/build#L83 the script continue and execute https://github.com/metabrainz/search-indexer/blob/d82fa5202836d2201de0e5b54a9c76fe559bc5b6/bin/build#L111 and following, rotating data
  - the same goes for https://github.com/metabrainz/search-indexer/blob/d82fa5202836d2201de0e5b54a9c76fe559bc5b6/bin/build#L111 which may rotate after pushing bad data

This is work-in-progress, but i would improve that stuff to be more reliable, and simpler.

Another thing that annoys me: the push/receive stuff is using multiple commands, i don't think it is needed.
Push does: sync/rotate/restart and receive has a prepare which looks unused.

I think we could simplify: use scp to copy data using "normal" ssh key (disabling compression/encryption), and on receiver side, just check for a new file to detect the end of the process

Something like (not actual code):

On sender:
 touch END ; scp -r data/cur/\* receiver:~/data/new/ && scp END receiver:~/data/END

On receiver:
 while true; do
   if END change : rotate new -> cur && restart search-server
   sleep 
end

**Comment by yvanzo** [comment]
Closing as this repository has been replaced with [SIR](https://github.com/metabrainz/sir) in June 2018, see the [blog post](https://blog.metabrainz.org/2018/06/15/musicbrainz-search-overhaul/).

**Pull request closed without merging**